File: src/web/magmac/app/compile/node/NodeList.ts
export interface NodeList {
	 iter() : Iter<Node>;
	 findLast() : Option<Node>;
	 add( element : Node) : NodeList;
	 addAll( others : NodeList) : NodeList;
	 join( delimiter : String,  generator : Function<Node, CompileResult<String>>) : CompileResult<String>;
}

File: packages/compiler-java/src/main/java/magmac/app/compile/node/NodeList.java
package magmac.app.compile.node;

import magmac.api.Option;
import magmac.api.iter.Iter;
import magmac.app.compile.error.CompileResult;

import java.util.function.Function;

/**
 * Immutable collection of {@link Node} values.
 */
public interface NodeList {
    Iter<Node> iter();

    Option<Node> findLast();

    NodeList add(Node element);

    NodeList addAll(NodeList others);

    CompileResult<String> join(String delimiter, Function<Node, CompileResult<String>> generator);
}
